typing/typedecl.ml

```diff
@@ -587,6 +586,12 @@ let check_coherence env loc dpath decl =
           end
       | _ -> raise(Error(loc, Definition_mismatch (ty, None)))
       end
+  | { type_kind = Type_abstract { immediate = imm };
+      type_manifest = Some ty } ->
+     begin match Ctype.check_type_immediate env ty imm with
+     | Ok () -> ()
+     | Error v -> raise(Error(loc, Immediacy v))
+     end
```

Comment:
There is something weird here:
- I don't know how you thought of adding this logic here. Probably from the testsuite? It does not fit the documentation of the function, so maybe it does not belong here, or maybe the description/documentation should be generalized.
- It is strange that you would check for immediacy errors in the Type_abstract case, but not in the other cases. (Is it not supported to annotate a constant-variants type with an immediacy requirement?) If this is not required because it is already done by `transl_declaration`, then could the abstract case also be checked there?

Reply:
This is definitely the right place for this check, but thanks for pointing out the doc comment I missed.

A type declaration has the general form `type _ t = M = K`, including both a manifest `M` (an optional type expression) and a kind `K` (one of abstract, a record definition, a variant definition, or `..` for open types). The `check_coherence` function checks that, when M is not None, it agrees with the kind, i.e:

  - if K is a variant type, check that M is a variant with the same parameters and constructors
  - if K is a record type, check that M is a record with the same parameters and fields
  - if K is `..`, check that M is an open type with the same parameters

Previously, `check_coherence` had no case for abstract types, because there was no way that an abstract type could be incoherent with any given `M`. Now, it is possible to write `Type_abstract {immediate=Immediate}` (i.e. an abstract type which abstracts only over the immediate types), and so there is now something worth checking in `check_coherence`.

> It is strange that you would check for immediacy errors in the Type_abstract case, but not in the other cases. (Is it not supported to annotate a constant-variants type with an immediacy requirement?) If this is not required because it is already done by transl_declaration, then could the abstract case also be checked there?

It is supported, and is checked as you say in transl_declaration. I don't think it would be correct to check the abstract case there, as the final environment isn't built yet.

I think the confusion in the current code is that transl_declaration actually does _too much_, because it tried to check the manifest as well as the kind. I've changed it to make the division of responsibility clearer:

  - transl_declaration ensures that the kind matches the immediacy declaration
  - check_coherence ensures that the kind and manifest are compatible